ASAN_TRAP | WebCore::LocalFrameViewLayoutContext::performLayout; WebCore::LocalFrameViewLayoutContext::layout; WebCore::Document::updateLayout

lopezlozoya opened the pull request:
#### 17237297720e4a207a4a63574c8902c33c6b3c61
<pre>
ASAN_TRAP | WebCore::LocalFrameViewLayoutContext::performLayout; WebCore::LocalFrameViewLayoutContext::layout; WebCore::Document::updateLayout
<a href="https://bugs.webkit.org/show_bug.cgi?id=293511">https://bugs.webkit.org/show_bug.cgi?id=293511</a>
<a href="https://rdar.apple.com/147938727">rdar://147938727</a>

Reviewed by NOBODY (OOPS!).

Splitting the fuction InspectorInstrumentation::didLayout in two parts, once to set the Quads,
and the other to did the layout, the later doesn&apos;t require the layoutRoot variable to exist.

* LayoutTests/fast/dom/StyleSheet/stylesheet-layout-insert-rule-audio-expected.txt: Added.
* LayoutTests/fast/dom/StyleSheet/stylesheet-layout-insert-rule-audio.html: Added.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::setLayoutQuadsImpl):
(WebCore::InspectorInstrumentation::didLayoutImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::setLayoutQuads):
(WebCore::InspectorInstrumentation::didLayout):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
(WebCore::PageTimelineAgent::setLayoutQuads):
(WebCore::PageTimelineAgent::didLayout):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::performLayout):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/17237297720e4a207a4a63574c8902c33c6b3c61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25203 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110692 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56161 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107533 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33744 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80121 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108498 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20060 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95205 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60430 "Passed tests") | 
| | [💥 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19806 "An unexpected error occured. Recent messages:Failed to print configuration") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13295 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55529 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89515 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113460 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32634 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24062 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89199 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88856 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33728 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11533 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/28091 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32560 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37973 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32316 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35663 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33903 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/InspectorInstrumentation.cpp:
```diff
@@ -528,10 +528,18 @@ void InspectorInstrumentation::willLayoutImpl(InstrumentingAgents& instrumenting
-        pageTimelineAgent->didLayout(root);
+        pageTimelineAgent->setLayoutQuads(root);
+    if (auto* pageAgent = instrumentingAgents.enabledPageAgent())
+        pageAgent->didLayout();
```
this shouldn't be here as otherwise it's already called by `InspectorInstrumentation::didLayout`

lopezlozoya replied:
Noted, I'll change it, thank you.

dcrousso commented on Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
```diff
@@ -286,10 +286,11 @@ void LocalFrameViewLayoutContext::performLayout(bool canDeferUpdateLayerPosition
         if (m_needsFullRepaint)
             renderView()->repaintRootContents();
         ASSERT(!layoutRoot->needsLayout());
+        InspectorInstrumentation::setLayoutQuads(frame, *layoutRoot);
         protectedView()->didLayout(layoutRoot, canDeferUpdateLayerPositions);
```
is it possible for the result of `RenderObject::absoluteQuads()` to change in between these two lines?

if so, this will be a difference in behavior and we likely don't want that

if not, i think we can solve this like Web Inspector has done for other similar things
```cpp
Vector<FloatQuad> layoutQuads;
/* ... */
{
    /* ... */
    if (InspectorInstrumentation::hasFrontends()) [[unlikely]]
        layoutQuads = layoutRoot->absoluteQuads(layoutQuads);
    protectedView()->didLayout(layoutRoot, canDeferUpdateLayerPositions);
    /* ... */
}
InspectorInstrumentation::didLayout(frame, layoutQuads);
/* ... */
```

lopezlozoya replied:
I am not really sure if that could change between those two lines, I am lack of experience to determine it. I'll try to analyze and check if that if possible.

alanbaradlay replied:
> is it possible for the result of RenderObject::absoluteQuads() to change in between these two lines?

yes. so what's happening here is actually normal and expected (though not very common)
At this point we are already in the 'post' phase of layout.
We've finished computing all the sizes and positions. However `protectedView()->didLayout` calls back to `FrameView` where we may find that the view needs space-taking scrollbars....which shrinks the available space for the content (see `availableContentSizeChanged`) and now we have to run another layout including re-resolving our style (still inside the previous `performLayout()`)
During this style resolve renderers may come and go (container query etc) and in this case, one of the renderers that go away is the previous layout's entry point, the layout root.
While this results in a recursive call back to `performLayout()`, technically we are done with the initial layout (no scrollbar) and this second one with the scrollbars on, is just a regular, subsequent layout .
if we moved `InspectorInstrumentation::didLayout` before `protectedView()->didLayout` InspectorInstrumentation would see a 
  [begin  end]
  [begin  end]
structure instead of what we current have which is 
 [begin [begin  end] end]

This would certainly be a behavior change though. @dcrousso what do you think?

dcrousso replied:
interesting!

Web Inspector really just uses the `RenderObject::absoluteQuads` to fill in the Area, Width, and Height columns of the Layout & Rendering timeline (as well as highlighting that area of the page when hovered)

i think the intention is to describe the area that did layout, so if the area can change after calling this then we probably want to save the value from before to use after

as far as moving the `InspectorInstrumentation`, im actually not sure whether we ever expect/allow `Timeline.EventType.Layout` to be nested inside other `Timeline.EventType.Layout`.  if the second layout is caused by the first, then it does make sense to have them be nested, so id err on the side of keeping that behavior as-is (i.e. the code i suggested above is probably what id do to fix this issue)

alanbaradlay replied:
@dcrousso makes sense, thank you. would you mind reviewing my patch at https://bugs.webkit.org/show_bug.cgi?id=293511?

dcrousso reviewed (commented):
yeah id also be very curious to understand how the `layoutRoot` is being invalidated as it makes me wonder if there's not other bugs elsewhere in this function (i.e. there's a lot of unchecked `->` and `*`)

dcrousso commented on Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
```diff
@@ -202,7 +202,15 @@ void PageTimelineAgent::didLayout(RenderObject& root)
+    if (!entry)
+        return;
+
+    ASSERT(entry->type == TimelineRecordType::Layout);
```
btw none of these lines are really needed as `InspectorTimelineAgent::didCompleteCurrentRecord` already performs these checks

lopezlozoya committed:
ASAN_TRAP | WebCore::LocalFrameViewLayoutContext::performLayout; WebCore::LocalFrameViewLayoutContext::layout; WebCore::Document::updateLayout
https://bugs.webkit.org/show_bug.cgi?id=293511
rdar://147938727

Reviewed by NOBODY (OOPS!).

Splitting the fuction InspectorInstrumentation::didLayout in two parts, once to set the Quads,
and the other to did the layout, the later doesn't require the layoutRoot variable to exist.

* LayoutTests/fast/dom/StyleSheet/stylesheet-layout-insert-rule-audio-expected.txt: Added.
* LayoutTests/fast/dom/StyleSheet/stylesheet-layout-insert-rule-audio.html: Added.
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::setLayoutQuadsImpl):
(WebCore::InspectorInstrumentation::didLayoutImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::setLayoutQuads):
(WebCore::InspectorInstrumentation::didLayout):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.cpp:
(WebCore::PageTimelineAgent::setLayoutQuads):
(WebCore::PageTimelineAgent::didLayout):
* Source/WebCore/inspector/agents/page/PageTimelineAgent.h:
* Source/WebCore/page/LocalFrameViewLayoutContext.cpp:
(WebCore::LocalFrameViewLayoutContext::performLayout):
